utils/exception_container.hh

```diff
@@ -89,6 +100,27 @@ public:
         return std::visit(std::move(f), *_eptr);
     }
 
+    // Explicitly clones the exception container.
+    // If the container was created on the current shard, then the new container
+    // will share the exception through the same lw_shared_ptr.
+    // However, if the container originates from a foreign shard, then a full
+    // clone will be performed.
+    //
+    // CAUTION: This method is _not_ noexcept if you clone a container
+    // which was created on another shard!
+    exception_container clone() const {
```

Comment:
Here is a small problem: I initially wrote the exception_container to be move only, but it turned out that there is a place in the code where I need to set the same exception to two promises (`_cl_promise` and `_done_promise`). It would be really nice if a class used for writing exception handling logic wouldn't throw on copy, so I considered the following options:

1. (Blasphemous, I know) Use std::shared_ptr. The main problem with it is that it uses an atomic for reference counting. However, this cost won't be paid on the happy path, and it can't be worse than std::exception_ptr because the latter is also copyable and thread-safe. In case when an exception _is_ returned using the exception_container, if the container is only moved around and never copied, then I think we will pay only one atomic operation per request (when the container is destroyed). In case of an exception storm, it can be a lot, though.

   Funnily enough, switching to std::shared_ptr makes the happy path faster by about 100 insns/op, probably because the container type is smaller (only a pointer, foreign_ptr holds pointer + shard ID).

2. Convert std::unique_ptr to lw_shared_ptr. If the container is copied on the same shard, then we only increase the reference count and that is cheap. However, if the container comes from another shard, then we perform a full copy, which involves an allocation. This requires adding a `copy_local` to seastar which I'm using here, I'll send a patch if we agree this is a good direction. This is my current approach, but I'm not sure what to do if the allocation fails (currently, copying can throw). I see two options:

	- Make copying noexcept and abort on allocation failure. This sounds hardcore, but that's what seastar does when it can't allocate a task. The exception container doesn't need much memory, only to hold a std::variant with an exception, so maybe it's not a problem.
	- Return an empty exception container. Currently, we treat an empty container as if it had a "bad access" exception inside, so maybe we can give a special meaning to it. Unfortunately, empty containers can also be created when a failed result which was moved out is accidentally returned.
	- Add an inline bool flag to exception_container that it failed to be constructed properly. This might increase the size of the exception container and make it more costly to move (I'll have to check).

Now that I wrote this, I realized that there already is a similar problem in the exception_container's constructor, which allocates and can fail.

Reviewers, what do you think? Which sounds the best?

Reply:
Since std::exception_ptr is morally equivalent to std::shared_ptr, we don't lose compared to status quo ante.

libstdc++ recently optimized the destructor of std::shared_ptr [1] for the case of non-shared data (which will be the common case in the uncommon path of having an exception), so I think it's fine. We'll take a hit until we upgrade, but without real sharing, the hit is very small. 

[1] https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=dbf8bd3c2f2cd2d27ca4f0fe379bd9490273c6d7

Reply:
While for writes the optimization you mentioned will make it work well, in case of read failures/timeouts the shared_ptr will have to be copied, see here: [link](https://github.com/scylladb/scylla/pull/10092/commits/2fde37b4c317efe3b6f822a901b25b9096a54f7e#diff-b973f4e30060a6f4beaf81347658f565fa895998c849d92e80f574a6255bad75L2858-R2862). So, for read errors, we will actually pay the price of atomics even with the optimization that you mentioned (although it will be 1 atomic decrement less).

What about the exception_container's constructor? It can fail because it allocates space to hold the exception. How should it behave in case of bad_alloc?

Reply:
I think it's okay.

For bad_alloc, it can store an std::exception_ptr containing a bad_alloc. We lose the timeout report, but if we can't allocate space for an exception, the system is dead anyway.

Reply:
I swtiched to std::shared_ptr and copying exception containers now does not throw.

As the problem of the throwing exception_container constructor is a bit orthogonal to this whole PR, I'll handle it in a followup.